packages/preview/src/browser/preview-contribution.ts

```diff
@@ -145,95 +123,72 @@ export class PreviewContribution implements CommandContribution, MenuContributio
     }
 
     protected registerOpenOnDoubleClick(previewWidget: PreviewWidget): void {
-        if (!previewWidget) {
-            return;
-        }
-        const disposable = previewWidget.onDidDoubleClick(location => {
-            const refWidget = this.findWidgetInMainAreaToAddAfter();
-            const widgetOptions: ApplicationShell.WidgetOptions = (refWidget)
-                ? { area: 'main', mode: 'tab-after', ref: refWidget }
-                : { area: 'main', mode: 'split-left' };
-            this.editorManager.open(new URI(location.uri), { widgetOptions })
-                .then(widget => {
-                    if (widget) {
-                        widget.editor.revealPosition(location.range.start);
-                        return widget.editor;
-                    }
-                }).then(editor => {
-                    if (editor) {
-                        editor.selection = location.range;
-                    }
-                });
+        const disposable = previewWidget.onDidDoubleClick(async location => {
+            const ref = this.findWidgetInMainAreaToAddAfter();
+            const { editor } = await this.editorManager.open(new URI(location.uri), {
+                widgetOptions: ref ?
+                    { area: 'main', mode: 'tab-after', ref } :
+                    { area: 'main', mode: 'split-left' }
+            });
+            editor.revealPosition(location.range.start);
+            editor.selection = location.range;
         });
         previewWidget.disposed.connect(() => disposable.dispose());
     }
 
-    canHandle(uri: URI, options?: OpenerOptions): number {
-        let canHandle = (this.previewHandlerProvider.canHandle(uri)) ? 50 : 0;
-        if (canHandle && uri.query.indexOf(this.id) >= 0) {
-            canHandle = 200;
+    async canHandle(uri: URI): Promise<number> {
+        if (!this.previewHandlerProvider.canHandle(uri)) {
+            return 0;
+        }
+        const editorPriority = await this.editorManager.canHandle(uri);
+        if (PreviewUri.match(uri)) {
+            return 2 * editorPriority;
         }
-        return canHandle;
+        return editorPriority / 2;
     }
 
-    async open(uri: URI, options?: PreviewOpenerOptions): Promise<PreviewWidget | undefined> {
-        try {
-            await this.resourceProvider(uri);
```

Comment:
that check is quite useful actually. in case of markdown preview, it shows you a warning about missing resource, instead of opening an empty and untitled preview, if the relative link is broken.

Reply:
The contract is that `OpenHandler.open` should reject if it cannot open. We should handle it on the client side.

Reply:
When does it happen? in `PreviewContribution.openForEditor`?

Reply:
In `PreviewContribution.open`; it happens in markdown preview, if a relative link is opened.

Reply:
Or do you mean, the caller should check, whether the resource exists, or not? In that case, the check should be actually moved to `MarkdownPreviewHandler`.

Reply:
Why does it get opened at all? Maybe we just should move looking up a resource out of the widget initialization then `open` will throw before a widget gets even created.

Reply:
That would be a better option. Then the `PreviewWidget.initialization` should be revised, as it is  called in widget's constructor now. Creation of preview widget depends on URI only, to make the widget restore work easily. 

I see your point.

Reply:
I've moved the resource initialization before the widget creation, but it is not enough since a file resource can exist even for an unexisting file, e.g. it is used to observer settings. We need something like `Resource.exists` and call it before creating a widget, the same should be done for editors. I would like to tackle it separately.